###snpden.R
###Developed by Henrique V Figueiro
###Contact: henriquevf@gmail.com
###Description: R script with the basic steps to generate a snp density plot based on the snpden function on vcftools

library(ggplot2)
library(gdata)
library(scales)
library(ggthemes)
library(ggsci)
library(svglite)
library(extrafont)

library(showtext)
font_add_google("Roboto Condensed")

showtext_auto() 
windows() 

setwd('working/directory')


snpden <- read.table('file.txt', header = T) #snpdensity table generated by vcftools

min(snpden$VARIANTS.KB)
max(snpden$VARIANTS.KB)
boxplot(snpden$VARIANTS.KB)

target <- c('HiC_scaffold_4','HiC_scaffold_16','HiC_scaffold_3',	#scaffolds you want to show in the plot, at this order
            'HiC_scaffold_1','HiC_scaffold_9','HiC_scaffold_17',
            'HiC_scaffold_7','HiC_scaffold_10','HiC_scaffold_6',
            'HiC_scaffold_5','HiC_scaffold_15','HiC_scaffold_19',
            'HiC_scaffold_2','HiC_scaffold_8','HiC_scaffold_18',
            'HiC_scaffold_14','HiC_scaffold_13','HiC_scaffold_12',
            'HiC_scaffold_11')

chr <-c('chr1','chr2','chr3','chr4','chr5','chr6','chr7','chr8','chr9','chr10',
        'chr11','chr12','chr13','chr14','chr15','chr16','chr17','chr18','chrX')  #chromosome names correspondent to the scaffolds 

chr.len <-c('0Mbp','50Mbp','100Mbp','150Mbp','200Mbp','250Mbp','300Mbp') #chromosome size 

snpden$CHROM <- reorder.factor(snpden$CHROM, new.order = target) #reorder table with target order 
snpden <-subset(snpden, snpden$CHROM!='NA') #remove rows without data

snpden$groups = cut(snpden$VARIANTS.KB, c(0,0.05,0.1,0.15,0.20,0.25,0.5,0.75,1,1.25,1.5,1.75,2,2.25,2.5,2.75,
                                          3,3.25,3.5,3.75,4,4.25,4.5,4.75,5)) #snpdensity intervals you wanna show

levels(snpden$CHROM) <-c('chr1','chr2','chr3','chr4','chr5','chr6','chr7','chr8','chr9','chr10',
                             'chr11','chr12','chr13','chr14','chr15','chr16','chr17','chr18','chrX')

title<-expression(paste(italic("Species name"), " sample heterozygous SNP densities"))

snpden_plot <- ggplot(data = snpden, aes(x=BIN_START, y=1)) + 
  geom_tile(aes(fill=groups)) +
  facet_grid(CHROM ~ ., switch='y') +
  xlab('Chromosome Length') + ylab('Scaffold Number') + theme_minimal(base_family='Roboto Condensed') +
  ggtitle(title) +
  theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major.y = element_blank(),
        strip.text.y.left = element_text(angle = 0, size=12),
        plot.title = element_text(lineheight = .8, hjust = .5, size = 14),
        plot.background = element_rect(fill = "white")) +
  scale_x_continuous(name='\nChromosome length', labels = c('0Mbp',"50Mbp", '100Mbp', "150Mbp", '200Mbp','250Mbp'),
                     breaks = c(0,50000000, 100000000, 150000000, 200000000,250000000)) +
  scale_fill_manual(values = c("#000081", "#0000f3", "#004dff", "#00b3ff", "#29ffce", 
                               "#7bff7b", "#ceff29", "#ffc600", "#ff6800", "#f30900", "#800000"),   #color scheme
                    name = "Variants/kb",
                    labels = c("<0.05","0.05-0.10","0.10-0.15","0.15-0.20","0.20-0.25",   #legend labels
                               "0.25-0.50","0.50-0.75","0.75-1.0","1.0-1.25","1.25-1.5",
                               "1.5-1.75"))


ggsave('sample.1Mb.snpden.svg', plot = snpden_plot, device = 'svg',
       dpi = 600, units = c('cm'), width = 26, height = 18)


ggsave('sample.1Mb.snpden.png', plot = snpden_plot, device = 'png',
       dpi = 600, units = c('cm'), width = 26, height = 18)
